README.md
<div align="center">
  <img src="https://raw.githubusercontent.com/cr0cK/bouchon/assets/assets/images/bouchon-icon.png" />
</div>

# bouchon

Efficient API mocking with cool libraries.

## Summary

- [Big picture](#big-picture)
- [Before using bouchon](#before-using-bouchon)
- [Quick start](#quick-start)
  - [JSON data](#json-data)
  - [Actions](#actions)
  - [Selectors](#selectors)
  - [Reducers](#reducers)
  - [Routes](#routes)
  - [Summary](#summary)
  - [Start bouchon](#start-bouchon)
- [Advanced usage](#advanced-usage)
  - [Export actions](#export-actions)
  - [Export selectors](#export-selectors)
  - [Multiple actions](#multiple-actions)
  - [Backend actions](#backend-actions)
  - [Asynchronous sample](#asynchronous-sample)
  - [Delays](#delays)
  - [Middlewares](#middlewares)
  - [Combine fixtures](#combine-fixtures)
  - [Use Babel for your fixtures](#use-babel-for-your-fixtures)
  - [Hot reload](#hot-reload)
- [Bouchon full API](#bouchon-full-api)
- [Use bouchon for integration testing](#use-bouchon-for-integration-testing)
- [Installation](#installation)
- [Other related packages](#other-related-packages)
- [License](#license)


## Big picture

Bouchon provides a way to make mocks easily with [redux](https://github.com/rackt/redux) and
[reselect](https://github.com/rackt/reselect).

Redux keeps your API stateful in order to create / edit / delete objects in a fake API and
reselect allows to retrieve any data from that state.

You define some data in a JSON file and your actions / reducers / selectors / middlewares / routes in a JS file.
These two files is what I call a _fixture_.

Each route (verb + url) defines an action, a selector and some optional middlewares.


## Before using bouchon

It is advisable to be comfortable with the following softs and techniques for using bouchon:

- [redux](https://github.com/rackt/redux)
- [reselect](https://github.com/rackt/reselect)
- [Currying techniques](http://www.sitepoint.com/currying-in-functional-javascript/)
- [ES2015](https://babeljs.io/docs/learn-es2015/)


## Quick start

Follow the following instructions to make your first fixture. The full code is available in
[the repository](https://github.com/cr0cK/bouchon/tree/master/tests/0-readme-tutorials).

Start an empty project, [install bouchon](#installation) and create folders and files:

```
$ mkdir bouchon-tutorial && cd $_
$ npm init
$ npm install bouchon babel-core babel-polyfill babel-preset-es2015 --save-dev
$ mkdir article && cd $_ && touch data.json && touch index.js
```

> Note: Examples are written in ES2015 and therefore, you have to install and configure Babel too.
> See the [Babel section](#use-babel-for-your-fixtures) for more informations.

### JSON data

Each fixture defines a JSON files with some data that will be saved in the state:

`// article/data.json`:

```
[
  {
    "id": 1,
    "title": "title 1",
    "body": "body 1"
  },
  {
    "id": 2,
    "title": "title 2",
    "body": "body 2"
  }
]
```

### Actions

In order to limit the boilerplate, bouchon uses [redux-act](https://github.com/pauldijou/redux-act)
that provides a simpler API to create actions and reducers. No types and switch cases are required.

Start by creating some actions:

```js
// article/index.js

import { createAction } from 'bouchon';

const actions = {
  get: createAction('Retrieve articles'),
  post: createAction('Create an article'),
  delete: createAction('Delete an article'),
};
```

### Selectors

Continue by writing some selectors. Selectors uses [reselect](https://github.com/rackt/reselect) and are
used to select data from the state.

```js
// article/index.js

import { createSelector } from 'bouchon';

// [...]

const selectors = {};
selectors.all = () => state => state.articles;
selectors.byId = ({id}) => createSelector(
  selectors.all(),
  articles => articles.filter(art => Number(art.id) === Number(id)).pop()
);
```

- The first curried function takes a merge of query, params and body parameters.
- The second takes the full `state` allowing to select any part of data.

### Reducers

Like every redux apps, you have to implement reducers that will maintain the state according to dispatched actions.

Here a basic implementation for the GET, POST, DELETE actions
(reminder, bouchon is using [redux-act](https://github.com/pauldijou/redux-act)
and the syntax is different that one you may know):

```js
// article/index.js

// [...]

const reducer = {
  [actions.get]: state => state,
  [actions.post]: (state, {body}) => ([
    ...state,
    body,
  ]),
  [actions.delete]: (state, {params}) => {
    const copy = state.slice(0);    // be careful to never mutate the state
    return copy.filter(art => Number(art.id) !== Number(params.id));
  },
};
```

Each function of the reducer takes the `state` the current fixture (not the full state)
and an object with `req`, `res`, `query` parameters, `params` parameters, `body` parameters
and must return a new state (be careful to *NEVER* mutate the state).

### Routes

Now, we have to define the routes of our API.

```js
// article/index.js

// [...]

const routes: {
  'GET /': {
    action: actions.get,
    selector: selectors.all,
    status: 200,
  },
  'GET /:id': {
    action: actions.get,
    selector: selectors.byId,
    status: 200,
  },
  'POST /:id': {
    action: actions.post,
    selector: selectors.byId,
    status: 201,
  },
  'DELETE /:id': {
    action: actions.delete,
    status: 204,
  },
};
```

### Summary

Add a `name` that will be used as the key in the state where your `data` live and you're done.

See the full fixture:

```js
// article/index.js

import { createAction, createSelector } from 'bouchon';

const actions = {
  get: createAction('Retrieve articles'),
  post: createAction('Create an article'),
  delete: createAction('Delete an article'),
};

const reducer = {
  [actions.get]: state => state,
  [actions.post]: (state, params) => ([
    ...state,
    params.body,
  ]),
  [actions.delete]: (state, {params}) => {
    const copy = state.slice(0);    // be careful to never mutate the state
    return copy.filter(art => Number(art.id) !== Number(params.id));
  },
};

const selectors = {};
selectors.all = () => state => state.articles;
selectors.byId = ({id}) => createSelector(
  selectors.all(),
  articles => articles.filter(art => Number(art.id) === Number(id)).pop()
);

const routes = {
  'GET /': {
    action: actions.get,
    selector: selectors.all,
    status: 200,
  },
  'GET /:id': {
    action: actions.get,
    selector: selectors.byId,
    status: 200,
  },
  'POST /:id': {
    action: actions.post,
    selector: selectors.byId,
    status: 201,
  },
  'DELETE /:id': {
    action: actions.delete,
    status: 204,
  },
};

export default {
  name: 'articles',
  data: require('./data.json'),
  reducer: reducer,
  endpoint: 'articles',
  routes: routes,
};
```

### Start bouchon

Bouchon is looking for files with `.fixture.js` suffix. Since your fixtures are
written in ES2015, you have to workaround a little bit in order to be able to load
them.

The tweak is just to create a unique file that requires all your fixtures. Just place
Babel at the top and you're done.

```js
// all.fixture.js

require('babel-core/register');
require('babel-polyfill');

module.exports = {
  default: [
    require('./article').default,
  ],
};
```

> Note: Don't forget to create your `.babelrc` file. See the [Babel section](#use-babel-for-your-fixtures) for more informations.

You can now start bouchon by providing your fixture folder and an optional port:

```
./node_modules/.bin/bouchon -d . -p 3000
```

You should see bouchon registering your urls:

```
00:32:21.938Z  INFO server: Registering "GET /articles/"
00:32:21.941Z  INFO server: Registering "GET /articles/:id"
00:32:21.942Z  INFO server: Registering "POST /articles/:id"
00:32:21.942Z  INFO server: Registering "DELETE /articles/:id"
00:32:22.001Z  INFO server: App listening at port 3000.
```

Do a GET to retrieve articles:

```bash
$ curl http://localhost:3000/articles  | python -mjson.tool
[
    {
        "body": "body 1",
        "id": 1,
        "title": "title 1"
    },
    {
        "body": "body 2",
        "id": 2,
        "title": "title 2"
    }
]
```

Do a DELETE to remove an article:

```bash
$ curl -X DELETE http://localhost:5556/articles/1
$ curl http://localhost:3000/articles  | python -mjson.tool
[
    {
        "body": "body 2",
        "id": 2,
        "title": "title 2"
    }
]
```

Hey, it works!


## Advanced usage

Now that you have an idea of how bouchon works, let's continue with complex workflows.

### Export actions

You should split your data and fixture by API namespace.
For example, if you have `./api/articles` and `./api/operations`, make two fixtures
that each handles their data.

But if you need to alter the state of an another fixture, you can export your
actions and reuse them in any fixture.

### Export selectors

In the same way, you can export selectors.

Imagine that you have 2 fixtures like articles and authors. If you want to join the data of both
in a single request, you can do something like that:

```js
// authors/index.js

export const selectors = {}
selectors.all = () => state => state.authors;

export default {
  name: 'authors',
  data: require('./data.json'),
};

// [...]

// articles/index.js

import { createSelector } from 'bouchon';
import { selectors as authorsSelectors } from '../authors';

export const selectors = {};
selectors.all = () => state => state.articles;

selectors.allWithAuthors = () => createSelector(
  selectors.all(),
  authorsSelectors.all(),
  (articles, authors) => {
    return articles.map(art => ({
      ...art,
      author: authors.filter(auth => Number(auth.id) === Number(art.authorId)).pop(),
    }));
  }
);

const routes = {
  'GET /': {
    selector: selectors.allWithAuthors,
    status: 200,
  },
};

export default {
  name: 'articles',
  data: require('./data.json'),
  routes: routes,
};
```

### Multiple actions

Because your data is splitted into several fixtures, you may need to alter the state
of several parts of your state for a single request.

For achieve that, you can use several actions for a route. Each action will be dispatched and will update
the state according to your reducers.

```js
const routes = {
  'POST /': {
    action: [articlesActions.create, operationsActions.setToDone],
    selector: selectors.byId,
    status: 200,
  },
};
```

### Backend actions

In some use cases, the workflow can be asynchronous. For example, an API request can just
respond "OK" and save an object in a queue processed later by a backend process.

Bouchon provides a feature named `backendAction` that allows to dispatch an action
in the future in order to simulate a backend process.

See how to return an object identifier in the request's response and create the object
five seconds later.

```js
const routes = {
  'POST /': {
    action: operationsActions.create,
    backendAction: {
      action: [actions.create, operationsActions.setToDone],
      delay: 5000,
    },
    selector: operationsSelectors.lastId,
    status: 201,
  },
};
```

### Asynchronous sample

To illustrate all the previous options, imagine that workflow:

- When I POST a request on `./articles`, it returns the detail of an operation
that contains the payload data and that will be handled later by an unknown process,
- when the operation has been processed, my article is created in the database
with the payload data of my operation,
- the operation is flagged as 'DONE',
- I can do a GET on `./articles` to list all created articles.

See how you can export / import actions and selectors and how to define your route
with `backendActions` in [this sample](https://github.com/cr0cK/bouchon/tree/master/tests/0-readme-tutorials/async-articles).


### Delays

Bouchon is useful to test your app in several conditions.

If you want to simulate a slow API, just add a delay in your `action`:

```js
const routes = {
  'GET /': {
    action: {action: operationsActions.create, delay: 2000},
    selector: selectors.all,
    status: 200,
  },
};
```

`delay` can also be an array like `[1000, 5000]` to simulate a delay between 1000 and 5000 milliseconds.


### Middlewares

You certainly know how Express middlewares work? bouchon supports middlewares in the same way.

For example, if you want a complete solution for pagination, you can easily write a middleware like this:

```js
// data is the selected data from your selector
const setPaginationHeaders = data => (req, res, next) => {
  const page = req.query.page || 1;
  const perPage = req.query.perPage || 10;
  const pageCount = Math.ceil(data.length / perPage);
  const totalCount = data.length;
  const slicedData = data.slice(
    (page - 1) * perPage,
    page * perPage
  );

  const headers = {
    'x-page': page,
    'x-per-page': perPage,
    'x-page-count': pageCount,
    'x-total-count': totalCount,
  };

  // if data are set in the response object, bouchon will return that data instead
  // of those selected by your selector
  res.data = slicedData;

  // set pagination headers
  res.set(headers);

  // do not forget to call next to continue the chain
  next();
};
```

Then just declare it in your route:

```js
const routes = {
  'GET /': {
    action: actions.get,
    selector: selectors.all,
    middlewares: [setPaginationHeaders],
    status: 200,
  },
};
```

### Combine fixtures

If you are already familiar with [redux](https://github.com/rackt/redux),
you certainly know how `combineReducers` work.

Bouchon is providing similar helpers to combine reducers and routes and thus allow
to split your fixtures and data by namespace by using the `endpoint` key.

It will combine the reducers and the routes.


```js
import { combineFixtures } from 'bouchon';

import books from './books';
import authors from './authors';

export default {
  name: 'library',
  endpoint: 'library',
  ...combineFixtures({
    books,
    authors,
  }),
};
```

### Use Babel for your fixtures

If you want to write your fixtures with modern Javascript, you have to workaround a
little bit in order to require Babel before your fixtures.

The simplest way is a have only [one fixture that requires all the others](https://github.com/cr0cK/bouchon/blob/master/tests/0-readme-tutorials/articles/all.fixture.js).

> Note: Don't forget to create a `.babelrc` at the root of your fixtures with the ES2015 preset:

```
{
  "presets": [ "es2015" ]
}
```

### Hot reload

Bouchon is providing a basic hot reload feature that allows to retrieve a previous state when the process is restarting.
It's useful when you just want to test a different status code / output in a route without losing all the context of your mocks.

To activate the feature, add the `--hot` option to your command line. When a query is performed, a `.bouchonHotReload` file
will be written with the full state. Bouchon will read this file when being restarted by a process supervisor like nodemon,
allowing to retrieve all the context of your mocks.

When the process is exited with `ctrl+c`, the file is removed.

Tips: To start Bouchon with nodemon, you can do something like this:

```
npm run start:dev -- -- -d /path/to/your/fixtures --hot
```


## Bouchon full API

TODO

## Use bouchon for integration testing

Bouchon is providing an API useful for integration tests.

For example, to test an app in a browser, you can start bouchon at the beginning of the test,
execute your test with a Selenium based tool and stop bouchon at the end.

Bonus: bouchon is recording all actions done during the test so you can check that
your process did exactly what you are expected at the end of your test.


```js
import path from 'path';
import chai from 'chai';
import freeport from 'freeport';
import request from 'request';
import { api as bouchon } from 'bouchon';


const expect = chai.expect;

describe('1 - List articles', function test() {
  this.timeout(10000);
  this.port = undefined;

  before((done) => {
    freeport((err, port) => {
      this.port = port;
      const pathFixtures = path.resolve(__dirname);
      bouchon.server.start({ path: pathFixtures, port })
        .then(() => done())
        .catch(done);
    });
  });

  after((done) => {
    expect(bouchon.logs.get()).to.deep.equal([{
      method: 'GET',
      originalUrl: '/articles',
      statusCode: 200,
      query: {},
      params: {},
      body: {},
    }]);

    bouchon.server.stop()
      .then(() => done())
      .catch(done);
  });

  it('should return articles', (done) => {
    request(`http://localhost:${this.port}/articles`, (err, res, body) => {
      if (err) { done(err); }

      expect(JSON.parse(body).length).to.equal(25);
      done();
    });
  });
});
```

## Installation

```
npm install --save bouchon
```

## Other related packages

- [bouchon-toolbox](https://github.com/cr0cK/bouchon-toolbox): a set of useful reducers, selectors and middlewares for common use cases
- [bouchon-samples](https://github.com/cr0cK/bouchon-samples): some examples for inspiration

## License

MIT.


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [2.0.0-alpha.4] - 2019-04-29
### Added
  - Add store in action meta. Allow to retrieve the store in reducers.

## [2.0.0-alpha.3]

## [2.0.0-alpha.2]

## [2.0.0-alpha.1]

## [0.4.0] - 2016-09-19
### Added
  - Hot reload feature (see README for more informations).

### Fixes
  - Eslint warnings.
  - Consistency in log calls.

## [0.3.3] - 2016-06-12
### Added
  - Save a reference to the store in the request object.
  - Add the possibility to dispatch custom actions (not handled by Bouchon).

## [0.3.2] - 2016-04-23
### Changed
  - Add babel tweak to README.

### Updated
  - Update dependencies versions.

## [0.3.1] - 2016-02-01
### Added
  - Add the support of meta data provided by [reduct-act](https://github.com/pauldijou/redux-act).

## [0.3.0] - 2016-01-23
### Changed
  - Ability to define several actions and backendAction for a route

```js
  'POST /': {
    action: [action.postArticle, action.postAuthor],
    status: 201,
  },
```

  - `responseBody` can be a function.

```js
  'POST /': {
    responseBody: args => state => { ... },
  },
```

  - Rewrite README.

## [0.2.1] - 2016-01-19
### Fixed
  - Params were not set in `backendAction`s.

## [0.2.0] - 2016-01-11
### Added
  - Reducers and routes are now combinable (new `combineFixturesReducers`, `combineFixturesRoutes`, `combineFixtures` functions).
  - Use `filterRows` and `extendRows` new implementation from bouchon-toolbox.
  - Expose `combineReducers` via bouchon (in case of you need it)

## [0.1.0] - 2015-12-30
### Added
  - `responseBody` key to quickly return arbitrary data instead of data of the state.
  - `backendAction` key to dispatch an action in the future in order to simulate asynchronous processes.

```js
  'POST /': {
    responseBody: { status: 'OK' },
    backendAction: {action: actions.postBackend, delay: 2000},
    status: 201,
  },
```

### Changed
  - `delay` key is deprecated, use the object notation for `action` to set a delay.
  - `action` and `backendAction` can now be a function or an object that defines a function and a delay.

```js
  'GET /:id': {
    action: {action: actions.get, delay: [400, 500]},
    selector: selectors.byId,
    status: 200,
  },
```

  - `action` key is optionnal. If not set, bouchon will emit a 'DUMMY_ACTION' (needed to enable logs).
  - `selector` key is optionnal.

## [0.0.2] - 2015-12-26
### Changed
  - Improve logger output.
  - Update README.

## 0.0.1 - 2015-12-16
  - Initial commit.


[Unreleased]: https://github.com/cr0cK/bouchon/compare/v0.4.0...HEAD
[0.4.0]: https://github.com/cr0cK/bouchon/compare/0.3.3...0.4.0
[0.3.3]: https://github.com/cr0cK/bouchon/compare/0.3.2...0.3.3
[0.3.2]: https://github.com/cr0cK/bouchon/compare/0.3.1...0.3.2
[0.3.1]: https://github.com/cr0cK/bouchon/compare/0.3.0...0.3.1
[0.3.0]: https://github.com/cr0cK/bouchon/compare/0.2.1...0.3.0
[0.2.1]: https://github.com/cr0cK/bouchon/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/cr0cK/bouchon/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/cr0cK/bouchon/compare/0.0.2...0.1.0
[0.0.2]: https://github.com/cr0cK/bouchon/compare/0.0.1...0.0.2


src/lib/router.js
/* eslint new-cap: 0 */
/* eslint arrow-body-style: 0 */

import fs from 'fs';
import path from 'path';
import express from 'express';
import _ from 'lodash';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { createAction } from 'redux-act';

import { outputLogger, activitiesLogger, hotReload } from '../middlewares/redux';
import { readState as readHotReloadState } from './hotReload';
import { concatEndpoint } from './combineFixturesRoutes';
import { combineFixturesReducers } from './combineFixturesReducers';
import { displayLogs } from './logger';


const router = express.Router();
const dummyAction = createAction('dummy_action');


/**
 * Check verb validity.
 *
 * @param  {String} verb
 * @return {Boolean}
 */
const checkVerbValidity = verb => {
  return ['get', 'post', 'put', 'patch', 'delete']
    .indexOf(verb.toLowerCase()) !== -1 ? verb : '';
};

/**
 * Create middlewares stack providing a `next` method like Express.
 *
 * @param  {Object} data  Selected data
 * @param  {Array} chain  List of middlewares
 * @return {Function}     First middleware to call
 */
const middlewareTrampoline = (data, chain) => {
  return function _trampoline(req, res, next) {
    function createNext(mdware, i) {
      return function _it(err) {
        if (err) {
          next(err);
          return;
        }

        const nextIndex = i + 1;
        const nextMiddleware = chain[nextIndex] ?
                               createNext(chain[nextIndex], nextIndex) :
                               next;
        try {
          mdware(data)(req, res, nextMiddleware);
        } catch (e) {
          next(e);
        }
      };
    }
    return createNext(chain[0], 0)();
  };
};

/**
 * Retreive all fixture files that match /\.fixture\.js$/
 * from a directory.
 *
 * @param  {String} fixturesDir   Directory of the fixtures
 * @return {Array}                List of paths
 */
const retrieveFixtures = fixturesDir => {
  const fixturesFound = [];

  const parseDir = dir => {
    fs.readdirSync(dir).forEach(file => {
      const fullDir = path.join(path.resolve(dir), file);
      const stats = fs.lstatSync(fullDir);
      if (stats.isDirectory()) {
        parseDir(fullDir);
        return;
      }
      if (stats.isFile() && /\.fixture\.js$/.test(file)) {
        fixturesFound.push(fullDir);
      }
    });
  };

  parseDir(fixturesDir);

  return fixturesFound;
};


/**
 * Load fixtures from an array of paths.
 * Each fixture must export a 'default' key.
 *
 * @param  {Array} fixturesFiles List of paths
 * @return {Array}               List of fixtures
 */
const loadFixtures = fixturesFiles => {
  return fixturesFiles.reduce((acc, fixturePath) => {
    try {
      let fixtureContent = require(fixturePath).default;  // eslint-disable-line global-require

      if (!_.isArray(fixtureContent)) {
        fixtureContent = [fixtureContent];
      }

      fixtureContent.forEach(fixtureContent_ => {
        if (!fixtureContent_.name) {
          console.error(`
Add a 'name' key in the specs of the fixture localised at ${fixturePath}.
It will be used to save fixture data in the store.
          `);
          return;
        }

        acc.push(fixtureContent_);
      });
    } catch (err) {
      console.error(`
Can't require the fixture from the path "${fixturePath}"
Cause: ${String(err)}
      `);
    }

    return acc;
  }, []);
};

/**
 * Retrieve all routes from fixtures and return an object of routes.
 *
 * @param  {Array} fixturesContent  Fixtures objects
 * @return {Objet}                  Routes (keys are 'VERB URL')
 */
const compileRoutes = fixturesContent => {
  return fixturesContent.reduce((acc, fixture) => {
    const routes = _.isObject(fixture.routes) ?
      concatEndpoint(fixture.routes, fixture.endpoint) :
      {};

    return Object.assign(acc, routes);
  }, {});
};

/**
 * Return the delay.
 * - Return the number if delay is a number.
 * - Return a delay between min and max if delay is an array
 * - Else 0
 *
 * @param  {Number | Array}
 * @return {Number}
 */
const getDelay = (delay) => {
  if (_.isNumber(delay)) {
    return delay;
  }

  if (_.isArray(delay)) {
    const [min, max] = delay;
    return Math.floor(Math.random() * (max - min + 1) + min);
  }

  return 0;
};

/**
 * Extract the delay and the action(s) function(s) from the action parameter
 * set in the route definition.
 *
 * If url is not defined, it does not fallback on a dummy action
 * (if action was not defined).
 *
 * @param  {Function | Array}
 * @return {Object}
 */
const extractActionParams = (action_, url) => {
  let delay = 0;
  let meta = null;
  let actions;

  // case: action: actions.get
  if (_.isFunction(action_)) {
    actions = [action_];
  // case: action: [actions.get, otherActions.get]
  } else if (_.isArray(action_)) {
    actions = action_;
  // case: action: {action: actions.get, delay: 500},
  // case: action: {action: [actions.get], delay: 500},
  } else if (_.isObject(action_)) {
    delay = getDelay(action_.delay);
    meta = action_.meta;
    actions = _.isArray(action_.action) ?
      action_.action : [action_.action];
  }

  // set a dummy action as fallback if url is set
  if (url) {
    if (_.isUndefined(action_)) {
      actions = [dummyAction];
    }

    if (_.isArray(actions)) {
      actions.forEach((action, i) => {
        if (!_.isFunction(action)) {
          console.error(`A action declared in the route "${url}" is not callable.`);
          actions[i] = dummyAction;
        }
      });
    }
  }

  return { actions, meta, delay };
};

/**
 * Return or call responsBody.
 *
 * @param  {Function | Mixed} responseBody
 * @param  {Object} selectorArgs
 */
const returnResponseBody = responseBody => selectorArgs => state => {
  if (_.isFunction(responseBody)) {
    return responseBody(selectorArgs)(state);
  }

  return responseBody;
};

/**
 * Initialize Redux store,
 * Define the router in charge of api routes.
 *
 * @param  {Object} options - path: Path of fixtures
 *                          - hot: Enable hot reload
 */
export const apiRouter = (options) => {
  const fixturesFiles = retrieveFixtures(options.path);
  const fixturesContent = loadFixtures(fixturesFiles);
  const allRoutes = compileRoutes(fixturesContent);
  const rootReducer = combineFixturesReducers(fixturesContent);

  // init store
  console.info('Initializing store...');

  const reduxMiddlewares = [thunk, outputLogger, activitiesLogger];
  if (options.hot) {
    reduxMiddlewares.push(hotReload);
  }

  let hotReloadPromise = Promise.resolve({});
  if (options.hot) {
    hotReloadPromise = readHotReloadState();
  }

  let store;
  hotReloadPromise.then((initialState) => {
    if (!_.isEmpty(initialState)) {
      console.info('✔ Hot-reloading from the latest known state.');
    }

    store = createStore(
      rootReducer,
      initialState,
      applyMiddleware(...reduxMiddlewares),
    );

    console.info('✔ Store is ready.');
  }).catch((err) => console.error(err));

  // register routes in the router
  _.forEach(allRoutes, (fixture, routeKey) => {
    const [verb, url] = routeKey.split(/\s+/);
    const method = checkVerbValidity(verb);
    const {
      action, backendAction,
      responseBody,
      selector, middlewares,
      status,
      delay,
    } = fixture;

    if (!method) {
      console.warn(`Verb "${verb}" is invalid! Skipping...`);
      return;
    }

    console.info(`Registering "${verb} ${url}"`);

    router[method.toLowerCase()](url, (req, res, next) => {
      if (!store) {
        console.error('Store is not ready, wait a second and try again...');
        next();
        return;
      }

      const actionParams = extractActionParams(action, url);
      const backendActionParams = extractActionParams(backendAction);
      const actionsArgs = [{
        query: req.query,
        // https://github.com/strongloop/express/issues/2734
        params: Object.assign({}, req.params),
        body: req.body,
        req,
        res,
        backendAction: backendActionParams,
        bouchonAction: true,
      }, actionParams.meta];

      // dispatch main actions
      if (_.isArray(actionParams.actions)) {
        actionParams.actions.map(action_ => store.dispatch(
          action_(...actionsArgs)
        ));
      }

      const selectorArgs = {
        ...req.query,
        ...req.params,
        ...req.body,
      };

      // get data from the selector
      let data;
      if (_.isFunction(selector)) {
        data = selector(selectorArgs)(store.getState());
      }

      // overwrite the status method to set a flag when setting the status (1)
      res.statusOriginal = res.status;  // eslint-disable-line no-param-reassign
      res.status = statusCode => {      // eslint-disable-line no-param-reassign
        res.statusDefined = true;       // eslint-disable-line no-param-reassign
        return res.statusOriginal(statusCode);
      };

      // get delay
      if (delay) {
        console.warn(
`The \`delay\` key is deprecated. Use this notation instead:
action: {action: myaction, delay: 1000}`);
      }

      // save the store in the request object, still usefull to retrieve
      // the state in some tricky use cases...
      req.store = store;                // eslint-disable-line no-param-reassign

      const actionDelay = delay || actionParams.delay;

      // return response
      setTimeout(() => {
        // apply middlewares
        if (_.isArray(middlewares)) {
          middlewareTrampoline(data, middlewares)(req, res, next);
        }

        // set status code and send data if not already done
        if (!res.headersSent) {
          // (1)
          if (status && !res.statusDefined) {
            res.statusOriginal(status);
          }

          // middlewares can save data in res.data to override the selected
          // data
          res.json(
            returnResponseBody(responseBody)(selectorArgs)(store.getState()) ||
            res.data ||
            data
          );

          next();
        }
      }, actionDelay);

      // dispatch backend actions
      if (_.isArray(backendActionParams.actions)) {
        setTimeout(() => {
          backendActionParams.actions.forEach(action_ => {
            // remove backendAction key to not add backend message in the
            // middleware
            delete actionsArgs[0].backendAction;
            store.dispatch(action_(...actionsArgs));
          });

          // display dispatched action(s) logs
          displayLogs(req.reduxLogs);

        // exec after 100 ms even if no delay defined in order to be called
        // after the action
        }, backendActionParams.delay || 100);
      }

      // when Bouchon is starting by a fork of the API process
      // (see api/server), the activities logs are sent to the child parent
      if ('send' in process) {
        process.send({ activityLog: req.activityLog });
      }
    });
  });

  return router;
};
